tempesta_fw/t/unit/test_frang.c

```diff
@@ -0,0 +1,389 @@
+#include <linux/slab.h>
+#include <linux/inet.h>
+#include <linux/module.h>
+#include <net/ipv6.h>
+#include <net/sock.h>
+#include <linux/socket.h>
+#include <linux/slab.h>
+
+#include "../../classifier.h"
+#include "../../client.h"
+#include "../../connection.h"
+#include "../../sync_socket.h"
+#include "../../gfsm.h"
+#include "../../http_msg.h"
+#include "../../log.h"
+#include "../../lib.h"
+#include "../../tempesta_fw.h"
+
+#include "addr.h"
+#include "cfg.h"
+#include "helpers.h"
+#include "kallsyms_helper.h"
+#include "sched_helper.h"
+#include "test.h"
+
+#define FRANG_HASH_BITS 17
+#define FRANG_FREQ	8
+
+ typedef struct {
+	unsigned long	ts;
+	unsigned int	conn_new;
+	unsigned int	req;
+} FrangRates;
+typedef struct frang_account_t {
+struct hlist_node hentry;
+struct in6_addr addr;		/* client address */
+unsigned long last_ts;	/* last access time */
+unsigned int conn_curr;	/* current connections number */
+FrangRates history[FRANG_FREQ];
+} FrangAcc;
+
+const int (*frang_conn_new) (struct sock *);
+struct sock mocksock;
+struct inet_sock *isk;
+int res;
+const char *inet_addr = "192.168.168.245.128";
+unsigned short i;
+typedef struct {
+	char *str;
+	size_t len;			/* The pre-computed strlen(@str). */
+} FrangCtVal;
+typedef struct {
+  /* Limits (zero means unlimited). */
+	unsigned int req_rate;
+	unsigned int req_burst;
+	unsigned int conn_rate;
+	unsigned int conn_burst;
+	unsigned int conn_max;
+/*
+   * Limits on time it takes to receive
+   * a full header or a body chunk.
+   */
+	unsigned long clnt_hdr_timeout;
+	unsigned long clnt_body_timeout;
+
+  /* Limits for HTTP request contents: uri, headers, body, etc. */
+	unsigned int http_uri_len;
+	unsigned int http_field_len;
+	unsigned int http_body_len;
+	unsigned int http_hchunk_cnt;
+	unsigned int http_bchunk_cnt;
+	bool http_ct_required;
+	bool http_host_required;
+/* The bitmask of allowed HTTP Method values. */
+	unsigned long http_methods_mask;
+/* The list of allowed Content-Type values. */
+	FrangCtVal *http_ct_vals;
+} FrangCfg;
+FrangCfg *frang_cfg;
+struct inet_sock *isk;
+
+TfwConnection *test_conn_alloc(void)
+{
+	static struct kmem_cache *test_conn_cache;
+	TfwConnection *conn;
+
+	if (!test_conn_cache) 
+		test_conn_cache = kmem_cache_create("tfw_test_conn_cache",
+					 sizeof(TfwConnection), 0, 0, NULL);
+	conn = kmem_cache_alloc(test_conn_cache, GFP_ATOMIC);
+	BUG_ON(!conn);
+	tfw_connection_init(conn);
+	return conn;
+}
+
+
+int
+req_handler(TfwHttpReq  *req)
+{
+	int (*frang_http_req_handler)(void *obj,
+				     struct sk_buff *skb, unsigned int off);
+
+	TfwConnection *conn;
+
+	conn = test_conn_alloc();
+	BUG_ON(!conn);
+	conn->msg = &req->msg;
+	conn->sk = &mocksock;
+	isk = (struct inet_sock *) (&mocksock);
+	isk->inet_saddr = htonl(in_aton(inet_addr));
+	if (!conn->sk->sk_security) {
+		frang_conn_new = get_sym_ptr("frang_conn_new");
+		res = frang_conn_new(conn->sk);
+	}
+
+
+
+	frang_http_req_handler = get_sym_ptr("frang_http_req_handler");
```

Comment:
frang_http_req_handler is local now. So getting it could not be moved to beginning TEST_SUIT. I did so for
frang_conf, frang_connn_new. For them it's Ok.